selene-lib/src/rules/mismatched_arg_count.rs

```diff
@@ -0,0 +1,425 @@
+
+        let mut visitor = MismatchedArgCountVisitor {
+            mismatched_arg_counts: Vec::new(),
+            scope_manager,
+            definitions,
+        };
+
+        visitor.visit_ast(&ast);
+
+        visitor
+            .mismatched_arg_counts
+            .iter()
+            .map(|mismatched_arg| {
+                Diagnostic::new_complete(
+                    "mismatched_arg_count",
+                    format!(
+                        "this function takes {}",
+                        mismatched_arg
+                            .parameter_count
+                            .to_message(mismatched_arg.num_provided)
+                    ),
+                    Label::new_with_message(
+                        mismatched_arg.call_range,
+                        mismatched_arg.parameter_count.to_string(),
+                    ),
+                    Vec::new(),
+                    vec![Label::new_with_message(
+                        mismatched_arg.function_definition_range,
+                        "note: function defined here".to_owned(),
+                    )],
+                )
+            })
+            .collect()
+    }
+
+    fn severity(&self) -> Severity {
+        Severity::Error
+    }
+
+    fn rule_type(&self) -> RuleType {
+        RuleType::Correctness
+    }
+}
+
+struct MismatchedArgCount {
+    parameter_count: ParameterCount,
+    num_provided: PassedArgumentCount,
+    call_range: (usize, usize),
+    function_definition_range: (usize, usize),
+}
+
+#[derive(Clone, Copy, Debug)]
+enum ParameterCount {
+    /// A fixed number of parameters are required: `function(a, b, c)`
+    Fixed(usize),
+    /// Some amount of fixed parameters are required, and the rest are variable: `function(a, b, ...)`
+    Minimum(usize),
+    /// A variable number of parameters can be provided: `function(...)`
+    Variable,
+}
+
+impl ParameterCount {
+    /// Calculates the number of required parameters that must be passed to a function
+    pub fn from_function_body(function_body: &ast::FunctionBody) -> Self {
+        let mut necessary_params = 0;
+
+        for parameter in function_body.parameters() {
+            match parameter {
+                ast::Parameter::Name(_) => necessary_params += 1,
+                ast::Parameter::Ellipse(_) => {
+                    if necessary_params == 0 {
+                        return Self::Variable;
+                    } else {
+                        return Self::Minimum(necessary_params);
+                    }
+                }
+            }
+        }
+
+        Self::Fixed(necessary_params)
+    }
+
+    /// Checks the provided number of arguments to see if it satisfies the number of arguments required
+    /// We will only lint an upper bound. If we have a function(a, b, c) and we call foo(a, b), this will
+    /// pass the lint, since the `nil` could be implicitly provided.
+    pub fn correct_num_args_provided(&self, provided: PassedArgumentCount) -> bool {
```

Comment:
No need for `&self` if this is `Copy`, then you get to ditch the `*required` and such as well.